enhancements/storage/vsphere-problem-detector.md

```diff
@@ -0,0 +1,230 @@
+
+## Proposal
+
+Run a dedicated vsphere-problem-detector operator that periodically tests OCP / vSphere:
+
+* The operator reports metrics, some of them sent via telemetry to us.
+* The operator reports alerts for common configuration issues.
+* The operator logs performed checks + more detailed info about OCP/vSphere config, so it's available in must-gather.
+
+Add pre-flight checks to installer.
+
+### Checks
+
+These checks will be performed regularly and / or in the installer:
+
+| Test | Periodic | Installer |
+|-|-|-|
+| OCP has enough permissions (e.g. list recent tasks and list datastores). | yes | yes |
+| Kubelets run with cloud-provider configured (i.e. have `Node.Spec.ProviderID`). |  yes | no |
+| All VMs have `disk.enableUUID=TRUE`. |  yes | no |
+| Datastore name used in OCP config is not too long. |  yes | yes |
+| Datastore names used in StorageClasses and PVs are not too long. |  yes | no |
+| Stretch goal: OCP can create a volume + attach it to a VM (this usually works when the checks above pass). |  yes | no |
+| Stretch goal: Was a VM / storage recently migrated using vMotion? We don't support this (yet). | yes | no |
+| Stretch goal: Is Storage Distributed Resource Scheduler enabled? We don't support this. | yes | no |
+
+When a previous test failed, the initial test frequency is once per 1 hour, doubling with every failed attempt up to once per 8 hours max.
+When a previous test succeeded, the frequency is once per 8 hours.
+
+### Metrics
+
+We want these metrics to be transferred via telemetry to us:
+
+* List of failed periodic tests.
+* HW version of vSphere VMs.
+* vCenter version.
+* ESXi host version (of each host).
+* List of installed storage plugins (3rd party vendor VIBs) - if possible.
+* List of enabled features:
+  * HA
+  * DRS
+  * SDRS w/DatastoreCluster
+
+
+### User Stories
+
+#### Story 1
+
+As a support person, I can see in must-gather that OCP is configured correctly on vSphere.
+
+I.e. results of the check(s) should be logged in must-gather.
+
+#### Story 2
+
+As OCP cluster admin, I want to know if I can create PVs and use them in pods. I.e. get an alert when something gets broken.
+
+#### Story 3
+
+As OCP PM, I want to know how many customers use an old vSphere version that may not be supported by a future OCP update and decide if I should push engineering harder to support the old vSphere version or convince few customers to upgrade.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+The operator will have direct access to vSphere API, which does not fine-grained permissions. As result, the permissions may be quite high. On the other hand, the operator exposes very little surface to attack.
+
+## Design Details
+
+All these are subject of change, see Open Questions below.
+
+* It is a standalone operator.
+* Running in namespace `openshift-cluster-storage-operator`.
+* Started by cluster-storage-operator (CSO) when it detects it runs on vSphere.
```

Comment:
Why is this a function of the storage operator?  How many vsphere problems have to do with storage vs api vs lb vs machines?

Reply:
Because so far we have only storage problems to check, see above. It would detect ~40%-50% of storage problems we have on vSphere, where users can make lot of mistakes.

Reply:
BTW, we can move it somewhere else if it's useful.